### ISSUE TITLE
No grad

### ISSUE THREAD

**Issue by richfwebb** [BY571/Upside-Down-Reinforcement-Learning#3]
In the inner for loop of algorithm 1 where you do sampling exploration, you don't want to accumulate gradients right, and can add 'with torch.no_grad()' ?

**Comment by BY571** [comment]
Thats true, i guess you could also just activate the evaluation mode of the behavior function. Havent tested it yet,  maybe it gives some improvements to the training time.

**Comment by richfwebb** [comment]
In the eval loop, should the intended horizon and reward be changing (like in the main algo loop)? I haven't checked the paper recently though to see what's right. One other idea, should an episode automatically terminate if horizon reaches <= 0, or intended reward <= 0 in the main algo?

**Comment by BY571** [comment]
What you are saying makes totally sense. I have to check the paper as well. I just did some test runs and with the new termination condition as you mentioned it converges slower...What makes sense since in the evaluation the behavior function is not forced to always get the highest reward for each step in the episode. So ill definitely check the paper again, thanks @richfwebb !

**Comment by BY571** [comment]
So the paper says 

> **at any time using evaluation commands derived from the most recent exploratory commands**. The initial desired return d_r is set to the lower bound of the desired returns from the most recent exploratory command, and the initial desired horizon d_h from the most recent exploratory command is reused.

even though they refer to algorithm 2 where they do decrease the commands. But it seems as if it's correct to not reduce reward and horizon, what do you think?

**Comment by richfwebb** [comment]
Yea the quote is probably saying initialize the params to that value, but still use algorithm 2 to update after each step. 
```
desired_return -= reward
desired_time_horizon -= 1
```
Just for someone else's take on it, I found this. They do update reward/horizon in evaluate_agent in [https://github.com/jscriptcoder/Upside-Down-Reinforcement-Learning/blob/master/Upside-Down_RL.ipynb](url)

They do have 2 additional parts, but they don't early terminate their episodes. I still need to check the paper. But here's what they add in that other github proj:

```
if not done and time_steps > max_steps:
            done = True
            reward = max_steps_reward

...
# Make sure it's always a valid horizon
desired_horizon = max(desired_horizon - 1, 1)
```

**Comment by BY571** [comment]
yeah its the implementation of a friend of mine :D 
You are right about that, thanks for pointing it out! I'll update it.

**Comment by richfwebb** [comment]
(From the other codebase) When bf.action is called from the evaluate function, with discrete actions, argmax is used instead of sampling. For continuous actions, do you think that should be done too? From the papers below, it seems ok to just return mu, because mode == mean for a normal distrib.

From the papers:
out(t) may encode the mean and the variance of a multidimensional Gaussian distribution over real-valued actions, from which a high-dimensional action out is sampled accordingly

2.3.6 Evaluation
		In certain conditions, greedy actions – using the mode of the action distribution – can also be used, but we omit this option here for simplicity

**Comment by BY571** [comment]
That's true, you can do that. 
Thanks, again for all the reviewing! How come you are so interested in Upside-down? 
I'm working on some improvements right now for the algorithm. Maybe you would be interested in connecting? If you want, join our group on discord https://discord.gg/mgJuh4

**Comment by richfwebb** [comment]
It says invite invalid right now.

**Comment by BY571** [comment]
check this one: https://discord.gg/Puvbhy

**Comment by richfwebb** [comment]
Third times a charm? I must have missed the notification, or richfwebb at gmail
